File: lsif/src/shared/models/hash.ts
import * as dumpModels from './dump'

/**
 * Hash a string or numeric identifier into the range `[0, maxIndex)`. The
 * hash algorithm here is similar to the one used in Java's String.hashCode.
 *
 * @param id The identifier to hash.
 * @param maxIndex The maximum of the range.
 */
export function hashKey(id: dumpModels.DefinitionReferenceResultId, maxIndex: number): number {
    const s = `${id}`

    let hash = 0
    for (let i = 0; i < s.length; i++) {
        const chr = s.charCodeAt(i)
        hash = (hash << 5) - hash + chr
        hash |= 0
    }

    // Hash value may be negative - must unset sign bit before modulus
    return Math.abs(hash) % maxIndex
}

File: enterprise/internal/codeintel/stores/lsifstore/hash.go
package lsifstore

// HashKey hashes a string identifier into the range [0, maxIndex)`. The
// hash algorithm here is similar ot the one used in Java's String.hashCode.
// This implementation is identical to the TypeScript version used before
// the port to Go so that we can continue to read old conversions without
// a migration.
func HashKey(id ID, maxIndex int) int {
	hash := int32(0)
	for _, c := range string(id) {
		hash = (hash << 5) - hash + c
	}

	if hash < 0 {
		hash = -hash
	}

	return int(hash % int32(maxIndex))
}
